File: packages/core/src/pacing.js
// Speech pacing — extracted verbatim from the Chrome POC.
// Gentle catch-up: when the translated sentence is clearly longer than its
// segment, speed up only A LITTLE — never more than +25% over the setting,
// nor ×1.45 absolute — then follow the player's own speed.

export const WORDS_PER_SECOND = 2.6; // at rate 1

// Space-less scripts (CJK, Thai): splitting on whitespace sees ONE "word"
// in a whole sentence, so the catch-up branch never fires and long lines
// get dropped downstream. Estimate words from character count instead.
const SPACELESS_RE =
  /[\u3040-\u30ff\u3400-\u4dbf\u4e00-\u9fff\uf900-\ufaff\uac00-\ud7af\u0e00-\u0e7f]/;

export function computeUtteranceRate({
  text,
  cueDur,
  baseRate,
  playbackRate = 1,
}) {
  const s = String(text || "").trim();
  const base = Number.isFinite(baseRate) && baseRate > 0 ? baseRate : 1;
  const words = SPACELESS_RE.test(s)
    ? Math.max(1, Math.round(s.replace(/\s+/g, "").length / 2.5))
    : s.split(/\s+/).filter(Boolean).length;
  const estimated = words / WORDS_PER_SECOND;
  let rate = base;
  if (cueDur > 0.5) {
    const ratio = estimated / base / cueDur;
    if (ratio > 1.15) {
      rate = Math.min(base * ratio, base * 1.25, 1.45);
    }
  }
  const pr = Number.isFinite(playbackRate) && playbackRate > 0 ? playbackRate : 1;
  return Math.min(rate * pr, 3);
}

File: apps/macos/VoxylioKit/Sources/VoxylioKit/Pacing.swift
// Speech pacing — port of packages/core/src/pacing.js. Gentle catch-up:
// never more than +25% over the setting, nor ×1.45 absolute, then follow
// the player's own speed. Arithmetic order mirrors JS for bit parity.

import Foundation

public let wordsPerSecond = 2.6 // at rate 1

/// Mirrors JS `text.split(/\s+/).length`: a leading or trailing
/// whitespace run yields an (empty) segment, and "" counts as one.
func jsWhitespaceSegmentCount(_ s: String) -> Int {
    var runs = 0
    var inRun = false
    for ch in s {
        if ch.isWhitespace {
            if !inRun {
                runs += 1
                inRun = true
            }
        } else {
            inRun = false
        }
    }
    return runs + 1
}

public func computeUtteranceRate(
    text: String, cueDur: Double, baseRate: Double, playbackRate: Double = 1
) -> Double {
    let words = Double(jsWhitespaceSegmentCount(text))
    let estimated = words / wordsPerSecond
    var rate = baseRate
    if cueDur > 0.5 {
        let ratio = estimated / baseRate / cueDur
        if ratio > 1.15 {
            rate = min(baseRate * ratio, baseRate * 1.25, 1.45)
        }
    }
    let pr = playbackRate == 0 ? 1 : playbackRate
    return min(rate * pr, 3)
}
